File: axo_endpoint/core/functions/params_validation.py
from __future__ import annotations

from typing import Any, Dict, List

from option import Err, Ok, Result

from axo_endpoint.core.errors import InvalidFieldError
from axo_shared.functions.params_schema import ParamSpec


def _type_matches(value: Any, param_type: str) -> bool:
    if param_type == "string":
        return isinstance(value, str)
    if param_type == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if param_type == "boolean":
        return isinstance(value, bool)
    if param_type == "json":
        return True
    if param_type == "data_ref":
        return isinstance(value, dict) and isinstance(value.get("kind"), str) and isinstance(value.get("location"), str)
    return False


def validate_and_fill_params(
    schema: List[ParamSpec], params: Dict[str, Any]
) -> Result[Dict[str, Any], InvalidFieldError]:
    """Validates a JOB_SUBMIT params dict against a function's declared
    params_schema: fills in defaults for missing optional fields, type-checks
    every declared field that's present, and rejects both missing-required
    fields and any key not declared in the schema (a strict whitelist once a
    schema exists). Callers must skip this entirely for an empty schema --
    that's the "accept an arbitrary blob" case, preserved for functions
    registered before params_schema existed."""
    declared_names = {spec.name for spec in schema}
    unknown = set(params.keys()) - declared_names
    if unknown:
        return Err(InvalidFieldError(
            f"unknown params not declared in params_schema: {sorted(unknown)}",
            context={"unknown_fields": sorted(unknown)},
        ))

    filled: Dict[str, Any] = {}
    for spec in schema:
        if spec.name not in params:
            if spec.required and spec.default is None:
                return Err(InvalidFieldError(
                    f"missing required param '{spec.name}'",
                    context={"field": spec.name},
                ))
            filled[spec.name] = spec.default
            continue

        value = params[spec.name]
        if not _type_matches(value, spec.type):
            return Err(InvalidFieldError(
                f"param '{spec.name}' does not match declared type '{spec.type}'",
                context={"field": spec.name, "expected_type": spec.type},
            ))
        filled[spec.name] = value

    return Ok(filled)

File: axo-ui/app/utils/paramsValidation.ts
import type { ParamSpec } from '~/types/functionRecord'

export interface ParamValidationError {
  field: string
  message: string
}

export type ParamValidationResult =
  | { ok: true, values: Record<string, unknown> }
  | { ok: false, errors: ParamValidationError[] }

function typeMatches(value: unknown, type: ParamSpec['type']): boolean {
  if (type === 'string') return typeof value === 'string'
  if (type === 'number') return typeof value === 'number' && Number.isFinite(value)
  if (type === 'boolean') return typeof value === 'boolean'
  if (type === 'json') return true
  if (type === 'data_ref') {
    return typeof value === 'object' && value !== null && !Array.isArray(value)
      && typeof (value as any).kind === 'string' && typeof (value as any).location === 'string'
  }
  return false
}

// Mirrors axo_endpoint/core/functions/params_validation.py::validate_and_fill_params.
// Fail-fast UX only -- the server re-validates unconditionally.
export function validateAndFillParams(
  schema: ParamSpec[],
  params: Record<string, unknown>,
): ParamValidationResult {
  if (schema.length === 0) return { ok: true, values: params }

  const declaredNames = new Set(schema.map(spec => spec.name))
  const unknown = Object.keys(params).filter(key => !declaredNames.has(key))
  if (unknown.length > 0) {
    return {
      ok: false,
      errors: unknown.map(field => ({ field, message: 'unknown param, not declared in params_schema' })),
    }
  }

  const errors: ParamValidationError[] = []
  const values: Record<string, unknown> = {}

  for (const spec of schema) {
    if (!(spec.name in params)) {
      const hasDefault = spec.default !== null && spec.default !== undefined
      if (spec.required && !hasDefault) {
        errors.push({ field: spec.name, message: `missing required param '${spec.name}'` })
        continue
      }
      values[spec.name] = spec.default ?? null
      continue
    }
    const value = params[spec.name]
    if (!typeMatches(value, spec.type)) {
      errors.push({ field: spec.name, message: `does not match declared type '${spec.type}'` })
      continue
    }
    values[spec.name] = value
  }

  return errors.length > 0 ? { ok: false, errors } : { ok: true, values }
}
